File: Python/LeetCodeTest/lc019_RemoveNthNodeFromEndOfList.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        last_node = head
        node_count = 1
        while last_node.next is not None:
            last_node = last_node.next
            node_count += 1
        if n == node_count:
            return head.next
        remove_id = node_count - n
        last_node = head
        result = last_node
        node_count = 1
        while last_node is not None and last_node.next is not None:
            if node_count == remove_id:
                last_node.next = last_node.next.next
            last_node = last_node.next
            node_count += 1
        return result

if __name__ == "__main__":
    solution = Solution()
    node = ListNode(1)
    node.next = ListNode(2)
    node.next.next = ListNode(3)
    node.next.next.next = ListNode(4)
    node.next.next.next.next = ListNode(5)
    result = solution.removeNthFromEnd(node, 5)
    while result is not None:
        print(result.val)
        result = result.next

File: Java/src/com/liu/LeetCode/CodeCaprice/LinkedList/LC019_RemoveNthNodeFromEndOfList.java
package com.liu.LeetCode.CodeCaprice.LinkedList;

import com.liu.LeetCode.common.bean.ListNode;
import com.liu.LeetCode.common.factory.NodeCreator;
import com.liu.LeetCode.common.tools.Utils;

public class LC019_RemoveNthNodeFromEndOfList {

    private static ListNode removeNthFromEndMy1(ListNode head, int n) {
        int length = 0;
        ListNode cur = head;
        while (cur != null) {
            cur = cur.next;
            length++;
        }
        ListNode pre = null; // 虚拟头节点
        cur = head;
        while (length > n) {
            pre = cur;
            cur = cur.next;
            length--;
        }
        if (cur == head) {
            return head.next;
        } else {
            pre.next = cur.next;
        }
        return head;
    }

    public static void main(String[] args) {
        int[][] valuesSet = new int[][]{
                {1, 2, 3, 4, 5},
                {1},
                {1, 2},
        };
        int[] nSet = new int[]{
                2,
                1,
                1,
        };
        ListNode result;
        long startTime;
        long endTime;

        for (int i = 0; i < valuesSet.length; i++) {
            ListNode node = (ListNode) NodeCreator.createNode(Utils.NODE_TYPE.NODE_TYPE_NORMAL_LIST, valuesSet[i]);
            startTime = System.currentTimeMillis();
            result = removeNthFromEndMy1(node, nSet[i]);
            endTime = System.currentTimeMillis();
            System.out.println("removeNthFromEnd My1 result" + i + " " + Utils.getNodePrintableLog(Utils.NODE_TYPE.NODE_TYPE_NORMAL_LIST,
                    result) + " during time " + (endTime - startTime));
        }
    }
}
